### DISCUSSION TITLE
谷歌翻译的 IP 好像又不能用了（2023 年 1 月 11 日）

### DISCUSSION THREAD

**Discussion by lmk123** [hcfyapp/crx-selection-translate#1607] [讨论]
今天（2023 年 1 月 11 日）谷歌翻译的可用 IP 似乎都失效了，且我使用 GoogleTranslate_IPFinder 扫描出来的 IP 实际上也是不能用的。GoogleTranslateIpCheck 还没试过。

2023 年 1 月 12 日更新：

应该只是部分 IP 失效了，见 https://github.com/hcfyapp/crx-selection-translate/discussions/1607#discussioncomment-4663327

## 一些背景

自 2022 年 9 月 28 日谷歌翻译退出中国市场后，网络上出现了第一批可以用的  IP。

之后在 10 月 20 日左右，第一批 IP 都失效了，还好 @kiird 提供了第二批可用 IP：https://github.com/hcfyapp/crx-selection-translate/discussions/1526#discussioncomment-3956157

再后来，@Ponderfly 和 @GoodCoder666 分别开发了集扫描可用 IP、测速、写入 hosts 为一体的软件：

- https://github.com/Ponderfly/GoogleTranslateIpCheck
- https://github.com/GoodCoder666/GoogleTranslate_IPFinder

但是，今天（2023 年 1 月 11 日）第二批 IP 似乎也失效了，且我使用 GoogleTranslate_IPFinder 扫描出来的 IP 实际上也是不能用的。GoogleTranslateIpCheck 还没试过。

**Comment by momo708** [comment]
确实已经不能用了
好像是全网IP都不能用了，我今天找了好多，都不能用了

**Comment by Loukky** [comment]
我这里测试美国的IP都能用呀,之前我一直用的新加坡的那个IP, 142.250.4.90
这个已经被阻断了,其他基本上都是正常的

**Reply by lmk123** [reply]
确实，这个 IP 在 IP 库第一个，所以导致不能用了。

**Comment by momo708** [comment]
142.251.117.90 translate.googleapis.com
这个IP，在国内内网就可用，不用开梯子就可以。目前唯一幸存IP

**Reply by Loukky** [reply]
这个是美国的吧,美国的IP我联通基本上都能用,另外新加坡还有几个IP也能用.联通到新加坡都是直连延迟只有60+

**Reply by lmk123** [reply]
我刚才发现 142.251.10.138 也能用

**Reply by lmk123** [reply]
自动化工具测出来的不准，在工具里 142.250.4.90 还是能用且最快的 :joy:

我准备挨个试一下看看哪些能用

**Reply by Loukky** [reply]
> 自动化工具测出来的不准，在工具里 142.250.4.90 还是能用且最快的 😂
> 
> 我准备挨个试一下看看哪些能用

这个IP,我联通已经在广州出口被阻断了

**Reply by lmk123** [reply]
142.250.4.90 我这边也用不了了，142.251.10.138 这个不是 90 结尾的你可以试试看能不能用

另外你是怎么确认一个 IP 是在哪里被阻断的？有相关的文档啥的吗

**Reply by Loukky** [reply]
> 142.250.4.90 我这边也用不了了，142.251.10.138 这个不是 90 结尾的你可以试试看能不能用
> 
> 另外你是怎么确认一个 IP 是在哪里被阻断的？有相关的文档啥的吗

142.251.10.138 我这不能用,直接在城域网就被阻断了.你跑个trace或者MTR就能看到是哪里断了的.现在防火墙镜像设备直接是下沉到市级机房的,很多阻断就直接是当地就给你阻断了

**Reply by Divem** [reply]
非常感谢，亲测有效！！！

**Comment by momo708** [comment]
108.177.97.100
108.177.111.90
108.177.122.90
108.177.125.186
108.177.126.90
108.177.127.90
142.250.0.90
142.250.1.90
142.250.4.90
142.250.8.90
142.250.9.90
142.250.10.90
142.250.11.90
142.250.12.90
142.250.13.90
142.250.27.90
142.250.28.90
142.250.30.90
142.250.31.90
142.250.96.90
142.250.97.90
142.250.98.90
142.250.99.90
142.250.100.90
142.250.101.90
142.250.102.90
142.250.103.90
142.250.105.90
142.250.107.90
142.250.111.90
142.250.112.90
142.250.113.90
142.250.114.90
142.250.115.90
142.250.123.90
142.250.125.90
142.250.126.90
142.250.128.90
142.250.138.90
142.250.141.90
142.250.142.90
142.250.145.90
142.250.152.90
142.250.153.90
142.250.157.90
142.250.157.183
142.250.157.184
142.250.157.186
142.250.158.90
142.250.159.90
142.251.1.90
142.251.2.90
142.251.4.90
142.251.5.90
142.251.6.90
142.251.8.90
142.251.9.90
142.251.10.90
142.251.12.90
142.251.15.90
142.251.16.90
142.251.18.90
142.251.107.90
142.251.111.90
142.251.112.90
142.251.116.90
142.251.117.90
142.251.120.90
142.251.160.90
142.251.161.90
142.251.162.90
142.251.163.90
142.251.166.90
172.217.192.90
172.217.195.90
172.217.203.90
172.217.204.90
172.217.214.90
172.217.215.90
172.253.58.90
172.253.62.90
172.253.63.90
172.253.112.90
172.253.113.90
172.253.114.90
172.253.115.90
172.253.116.90
172.253.117.90
172.253.118.90
172.253.119.90
172.253.123.90
172.253.124.90
172.253.125.90
172.253.126.90
172.253.127.90
216.58.227.65
216.58.227.66
216.58.227.67
我原来直接全部复制进去的，都不能用了
目前只有这个幸存142.251.117.90 translate.googleapis.com（不架梯子VPN就可以使用）

发件人：Loukky ***@***.***>
发送日期：2023-01-11 14:16:51
收件人：hcfyapp/crx-selection-translate ***@***.***>
抄送人：momo708 ***@***.***>,Comment ***@***.***>
主题：Re: [hcfyapp/crx-selection-translate] 谷歌翻译的 IP 好像又不能用了（2023 年 1 月 11 日） (Discussion #1607)

我这里测试美国的IP都能用呀,之前我一直用的新加坡的那个IP,那个已经被阻断了

**Reply by Loukky** [reply]
其中尾数是90的IP,我这边基本上都能用

**Reply by Loukky** [reply]
另外我测试了下这个彰化的IP我这也能用,即 108.177.97.100

**Comment by GoodCoder666** [comment]


**Reply by lmk123** [reply]
给力

**Reply by Loukky** [reply]
142.250.4.160/161/162/164 这个4个IP,新加坡的,联通直连只有差不多60ms.
另外如果测试模拟访问的话,可以在工具里面使用tcping的方式测80/443能不能通就可以反映出来能不能直接用了

**Reply by GoodCoder666** [reply]
> 142.250.4.160/161/162/164 这个4个IP,新加坡的,联通直连只有差不多60ms.

好像移动不怎么给力啊 😢 

> 另外如果测试模拟访问的话,可以在工具里面使用tcping的方式测80/443能不能通就可以反映出来能不能直接用了

我一开始就是这么想的，然而这样不行，会扫出来一堆用不了的IP。具体来说目前是发送一个测试包，看有没有返回正确结果，如果出现404或请求超时即视为不可用。

**Reply by Loukky** [reply]
> > 142.250.4.160/161/162/164 这个4个IP,新加坡的,联通直连只有差不多60ms.
> 
> 好像移动不怎么给力啊 😢
> 
> > 另外如果测试模拟访问的话,可以在工具里面使用tcping的方式测80/443能不能通就可以反映出来能不能直接用了
> 
> 我一开始就是这么想的，然而这样不行，会扫出来一堆用不了的IP。具体来说目前是发送一个测试包，看有没有返回正确结果，如果出现404或请求超时即视为不可用。

你如果host头只有域名就会发现提示的都是404,带一个资源文件就不会404了.比如 https://translate.googleapis.com/translate_static/css/translateelement.css
另外扫出来一堆用不了的ip是因为谷歌的api接口的CDN与谷歌其他产品的CDN不是通用的.所以会不能用
比如 storage.googleapis.com 这个域名直接是解析到谷歌台湾的机房但是防火墙没有屏蔽.
刚刚新加坡那几个IP我重新测试了下.我这80通,但是443不通,所以也没法用.毕竟谷歌的域名基本上都上https了.可能443端口被阻断了,老实说真的挺可惜.联通跟谷歌有直接peer全是到香港就接入谷歌直接的线路,延迟很低.如果新加坡没有可用的IP的话,就只能用美国的IP,延迟很高
所以这个扫描结果不正确
![](https://imgs3.loukky.com/imgs/2023/01/11/c43fbb7342b1fcda.png)

**Reply by GoodCoder666** [reply]
> 你如果host头只有域名就会发现提示的都是404,带一个资源文件就不会404了.比如 https://translate.googleapis.com/translate_static/css/translateelement.css 另外扫出来一堆用不了的ip是因为谷歌的api接口的CDN与谷歌其他产品的CDN不是通用的.所以会不能用 比如 storage.googleapis.com 这个域名直接是解析到谷歌台湾的机房但是防火墙没有屏蔽. 刚刚新加坡那几个IP我重新测试了下.我这80通,但是443不通,所以也没法用.毕竟谷歌的域名基本上都上https了.可能443端口被阻断了,老实说真的挺可惜.联通跟谷歌有直接peer全是到香港就接入谷歌直接的线路,延迟很低.如果新加坡没有可用的IP的话,就只能用美国的IP,延迟很高 所以这个扫描结果不正确

感谢提醒，我注意到代码中的一个问题：扫描时默认用的HTTP，没有启用HTTPS。是不是应该改用HTTPS请求？

默认的请求接口格式如下：
```
http://{ip}/translate_a/single?client=gtx&sl=en&tl=fr&q=a
```

**Reply by Loukky** [reply]
> 
> 感谢提醒，我注意到代码中的一个问题：扫描时默认用的HTTP，没有启用HTTPS。是不是应该改用HTTPS请求？
> 
> 默认的请求接口格式如下：
> 
> ```
> http://{ip}/translate_a/single?client=gtx&sl=en&tl=fr&q=a
> ```

你这样直接改成https不行的吧,https必须要带host头不然会报证书错误的.
估计你这代码要修改很多地方
至于http有没有用老是说我也不太清楚,我主要是在Chrome里面翻译,Chrome应该是走的https.
毕竟刚刚那几个80通但443不通的Ip我这测试没办法翻译.
另外请求的链接可以用我刚刚发的那个,不是挺好的

**Reply by GoodCoder666** [reply]
> 你这样直接改成https不行的吧,https必须要带host头不然会报证书错误的.
估计你这代码要修改很多地方
至于http有没有用老是说我也不太清楚,我主要是在Chrome里面翻译,Chrome应该是走的https.
毕竟刚刚那几个80通但443不通的Ip我这测试没办法翻译

默认就会带host头，不带host头会404，我正在修改，稍等

**Reply by lmk123** [reply]
@GoodCoder666 期待你修复扫描出来的 IP 不能用的 bug，我一个个测也费劲 :joy:

修复之后，我就可以直接用你的软件扫一份 IP 出来，然后著明出处来自你的软件了

**Reply by GoodCoder666** [reply]
> @GoodCoder666 期待你修复扫描出来的 IP 不能用的 bug，我一个个测也费劲 😂
> 
> 修复之后，我就可以直接用你的软件扫一份 IP 出来，然后著明出处来自你的软件了

哈哈，感谢。

不过进展不太顺利，我一直遇到这个错误：
```
[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:997)
```

根据网上的方法，改用TLSv1 Protocol，结果成了：
```
[SSL: NO_CIPHERS_AVAILABLE] no ciphers available (_ssl.c:997)
```

不知道有没有人有办法解决？

> **Note**
> 根据报错信息，问题应该是API接口使用了SSLv3协议，但是这个协议被认为不安全，就被默认禁用了……
> 用同样的方法访问`https://www.baidu.com/`之类的都没错

**Reply by Loukky** [reply]
> > @GoodCoder666 期待你修复扫描出来的 IP 不能用的 bug，我一个个测也费劲 😂
> > 修复之后，我就可以直接用你的软件扫一份 IP 出来，然后著明出处来自你的软件了
> 
> 哈哈，感谢。
> 
> 不过进展不太顺利，我一直遇到这个错误：
> 
> ```
> [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:997)
> ```
> 
> 根据网上的方法，改用TLSv1 Protocol，结果成了：
> 
> ```
> [SSL: NO_CIPHERS_AVAILABLE] no ciphers available (_ssl.c:997)
> ```
> 
> 不知道有没有人有办法解决？
> 
> > Note
> > 根据报错信息，问题应该是API接口使用了SSLv3协议，但是这个协议被认为不安全，就被默认禁用了……
> > 用同样的方法访问`https://www.baidu.com/`之类的都没错

这个似乎是Python的锅，谷歌上好像有解决方案

**Reply by aeeq** [reply]
https://www.zhihu.com/question/266700601/answer/2263452766

**Reply by GoodCoder666** [reply]
@Loukky 非常奇怪，他用的似乎是TLS1.3，而不是错误提示的SSLv3

<img width="259" alt="image" src="https://user-images.githubusercontent.com/59763425/211780730-488eae85-a0a0-4042-95aa-b3545230fcf6.png">

**Reply by GoodCoder666** [reply]
> https://www.zhihu.com/question/266700601/answer/2263452766

@aeeq 感谢，原本用的就是这个方法，不过用的是python3自带的urllib，我改用requests试试

**Reply by GoodCoder666** [reply]
目前看来还是没用

**Reply by lmk123** [reply]
我不太熟 Python，但我用 Node.js 试了一下是可以的，你可以试一下是不是要设置一下 Host 请求头之类的。

Node.js 代码如下：

```js
import https from 'node:https'

function checkIP(ip, timeout = 10 * 1000) {
  return new Promise((resolve) => {
    const now = Date.now()
    const req = https.get(
      `https://${ip}/translate_a/element.js`,
      {
        timeout,
        headers: {
          Host: 'translate.googleapis.com',
        },
        servername: 'translate.googleapis.com',
      },
      (res) => {
        resolve({
          valid: res.statusCode === 200,
          time: Date.now() - now,
        })
        res.resume()
      }
    )

    req.on('timeout', () => {
      resolve({ valid: false, timeout: true })
      req.destroy()
    })

    req.on('error', (err) => {
      resolve({
        valid: false,
        error: err,
      })
    })
  })
}

checkIP('108.177.97.100').then(console.log)
```

**Reply by lmk123** [reply]
> 非常奇怪，他用的似乎是TLS1.3，而不是错误提示的SSLv3

有没有可能服务器用的是 TLS 1.3，但是你的 python 代码是用 SSL v3 请求的，所以导致协议不匹配之类的？

如果是的话，那是不是改一下请求时使用的 SSL 协议就行了，比如在 Node.js 里就是可以修改的：https://stackoverflow.com/a/13589194

**Reply by GoodCoder666** [reply]
> 有没有可能服务器用的是 TLS 1.3，但是你的 python 代码是用 SSL v3 请求的，所以导致协议不匹配之类的？
> 
> 如果是的话，那是不是改一下请求时使用的 SSL 协议就行了，比如在 Node.js 里就是可以修改的：https://stackoverflow.com/a/13589194

应该不是，Python 的文档里都写的是 SSLv3 Deprecated

**Comment by aeeq** [comment]
一些可用的 ipv6
2607:f8b0:4001:c07::5a
2404:6800:4008:c13::5a
2404:6800:4008:c15::5a
2404:6800:4003:c00::5a
2404:6800:4003:c01::5a
2404:6800:4003:c02::5a
2607:f8b0:4023:1402::5a
2607:f8b0:4023:1401::5a
2607:f8b0:4023:401::5a
2607:f8b0:4023:1::5a
2607:f8b0:4023::5a

**Reply by lmk123** [reply]
感谢提供，我会将这些地址更新到公告当中

**Reply by Loukky** [reply]
这个好,有好几个彰化的IP,都能直接用

**Comment by lmk123** [comment]
我写了个工具用于检测 IP 是否可用，然后发现原有的 IP 库中差不多四分之一失效了，但剩下的四分之三还是可以用的。不能用的 IP 我已经在 https://hcfy.app/blog/2022/09/28/ggg 里标注了。

**Comment by GoodCoder666** [comment]
> @lmk123 @Loukky 我分析了一下`urllib`的源码，问题根源找到了。

[工具](https://github.com/GoodCoder666/GoogleTranslate_IPFinder)已完全改用HTTPS：
- Release：https://github.com/GoodCoder666/GoogleTranslate_IPFinder/releases/tag/alpha
- Windows：https://github.com/GoodCoder666/GoogleTranslate_IPFinder/releases/download/alpha/checker-win-x64.exe
- Mac OS：https://github.com/GoodCoder666/GoogleTranslate_IPFinder/releases/download/alpha/checker-mac.zip

现在应该不会出现扫描出来用不了的问题了。另外，我还在扫描里边加了个指定`timeout`的选项，方便获取完整的 IP 列表。

**Reply by Loukky** [reply]
感谢，这样应该就比较贴近实际，也更准确了

**Reply by GoodCoder666** [reply]
好像`142.250.4.90`又能用了 😅

**Reply by Loukky** [reply]
> 好像`142.250.4.90`又能用了 😅

这个IP，是在运营商的出口被阻断，和其他谷歌IP不同，其他的IP是直接在城域网阻断，所以这个新加坡的IP有可能是临时阻断，也有可能是目前在进行全国同步

**Comment by lxfly2000** [comment]
我的今天也不能用了

**Reply by lmk123** [reply]
有试过我在两天前更新的新的 IP 吗？

**Reply by lxfly2000** [reply]
感谢更新，换了个IPv6的地址可以用了

**Comment by GoodCoder666** [comment]
大家可以试一试[新版](https://github.com/GoodCoder666/GoogleTranslate_IPFinder/releases/tag/alpha)，加了一个IPv6解析功能：

<img width="175" alt="image" src="https://user-images.githubusercontent.com/59763425/212523886-a53ecb2c-b093-40ee-b409-8e73a41845a7.png">

一般只能解析到一两条，后续会改进。

原理是对特定的 DNS 服务器（不是系统默认的）发送解析请求，获取 IPv6 地址。
使用 DoH（DNS over HTTPS）协议，防止[DNS污染](https://baike.baidu.com/item/DNS%E6%B1%A1%E6%9F%93/8620359)。

**Reply by Loukky** [reply]
doh能否搞个自定义？就是让自己来填写的。
或者也可以内置几个比较流行的DNS

**Reply by GoodCoder666** [reply]
> doh能否搞个自定义？就是让自己来填写的。 或者也可以内置几个比较流行的DNS

DoH 不太好自定义，因为各个的接口不一致，不过后续会内置更多的 DNS，将所有解析到的 IPv6 都试一遍。

不过这些都只是临时的解决方案，如果我们能找到谷歌的 IPv6 网络，就可以像 IPv4 一样进行完整的扫描了。

**Reply by Loukky** [reply]
> > doh能否搞个自定义？就是让自己来填写的。 或者也可以内置几个比较流行的DNS
> 
> DoH 不太好自定义，因为各个的接口不一致，不过后续会内置更多的 DNS，将所有解析到的 IPv6 都试一遍。
> 
> 不过这些都只是临时的解决方案，如果我们能找到谷歌的 IPv6 网络，就可以像 IPv4 一样进行完整的扫描了。

几个公共DNS的doh接口都是一样的，不过内置不方便的话可以整个自定义方便使用者手动填写

**Reply by GoodCoder666** [reply]
> 几个公共DNS的doh接口都是一样的，不过内置不方便的话可以整个自定义方便使用者手动填写

我今天看了Cloudflare和阿里的服务，Cloudflare是直接传网址，阿里是要先base64，像这种不太好自定义吧

**Reply by Loukky** [reply]
> > 几个公共DNS的doh接口都是一样的，不过内置不方便的话可以整个自定义方便使用者手动填写
> 
> 我今天看了Cloudflare和阿里的服务，Cloudflare是直接传网址，阿里是要先base64，像这种不太好自定义吧

我看了下阿里dns的相关文档,转base64的也有,不过你没有必要用那个方法呀.\
比如`https://dns.alidns.com/resolve?name=www.google.com`
你这个可以正常的拿到解析结果吧,这个是doh json api接口.
但是,doh的接口,公共dns一般是都是`https://域名://dns-query`,例如阿里dns:`https://dns.alidns.com/dns-query`或者cloudflare的dns:`https://cloudflare-dns.com/dns-query`,谷歌的dns:`https://dns.google.com/dns-query`.其中传递的相关参数都是有固定的规范来实现的.
好像是这个:
`https://www.rfc-editor.org/rfc/rfc8484`

**Comment by Eacls** [comment]
1月15日，上面那位老哥提供的ipv6我随机测了三个发现都不可用了，ipv4的还可以。

**Reply by GoodCoder666** [reply]
试一下用新版的工具扫描？

https://github.com/GoodCoder666/GoogleTranslate_IPFinder/releases/tag/alpha

**Comment by GoodCoder666** [comment]
更新了一个扩展扫描，会从更大的范围中扫描 IP（包括 IPv6），不知道好不好用。建议使用时选择200线程。

https://github.com/GoodCoder666/GoogleTranslate_IPFinder/releases/tag/alpha

**Reply by aeeq** [reply]
新版扫到低延迟IP，但是用不了，比如 216.239.32.21 translate.googleapis.com

**Comment by ghost** [comment]
试试这两个
34.147.100.198
34.90.145.142
或者用https://github.com/rabbit2123/domain-fronting

**Reply by lmk123** [reply]
感谢提供 IP，测试了下这俩 IP 是可以用的，我会把它们加入到可用 IP 列表当中

**Reply by lmk123** [reply]
这俩 IP 好像已经失效了 :joy:

**Reply by ghost** [reply]
被认证了
